### ISSUE TITLE
Assertion `dst->empty()' failed

### ISSUE THREAD

**Issue by mahmoodn** [accel-sim/accel-sim-framework#134]
Hi
For a given kernel, the simulation hit the following assertion
```
abstract_hardware_model.cc:114: void move_warp(warp_inst_t*&, warp_inst_t*&): Assertion `dst->empty()' failed.
```

According to the following section and the extra printfs
```
void pipelined_simd_unit::cycle() {
  printf("(1) active_insts_in_pipeline = %u\n", active_insts_in_pipeline);
  if (!m_pipeline_reg[0]->empty()) {
    m_result_port->move_in(m_pipeline_reg[0]);
    assert(active_insts_in_pipeline > 0);
    active_insts_in_pipeline--;
  }
  printf("(2) active_insts_in_pipeline = %u\n", active_insts_in_pipeline);
  if (active_insts_in_pipeline) {
    for (unsigned stage = 0; (stage + 1) < m_pipeline_depth; stage++) {
      printf("Moving warp for stage %u\n", stage);
      move_warp(m_pipeline_reg[stage], m_pipeline_reg[stage + 1]);
    }
  }
  if (!m_dispatch_reg->empty()) {
    if (!m_dispatch_reg->dispatch_delay()) {
      int start_stage =
          m_dispatch_reg->latency - m_dispatch_reg->initiation_interval;
      printf("start_stage = %u\n", start_stage);
      move_warp(m_pipeline_reg[start_stage], m_dispatch_reg);
      active_insts_in_pipeline++;
    }
  }
  occupied >>= 1;
}

```
the output looks like
```
(1) active_insts_in_pipeline = 2
(2) active_insts_in_pipeline = 1
Moving warp for stage 0
Moving warp for stage 1
Moving warp for stage 2
start_stage = 1
accel-sim.out: abstract_hardware_model.cc:114: void move_warp(warp_inst_t*&, warp_inst_t*&): Assertion `dst->empty()' failed.

```

It seems that the pipeline is full but there is no mechanism to handle the full condition since the assertion comes from "being not empty". I haven't tried but maybe tweaking `dispatch_delay()` is a workaround for that. But even if that is a quick workaround, it is not a permanent fix and maybe be seen in another kernel. Can someone confirm that? have you seen that before?

**Comment by JRPan** [comment]
May I ask what the config and application you are running are? 

Thanks

**Comment by mahmoodn** [comment]
I have missed the config information. I am searching for it to reproduce that. I will let you know once I catch it.

**Comment by JRPan** [comment]
Thank you but you don't have to. I don't want to waste your time on that. 
In the issue #171 someone reported the same error with a Pascal config. Just want to see if these can be related. It's okay if you don't remember. Forget about that

Thanks!

**Comment by mahmoodn** [comment]
Please see the attached test case. I was able to copy a subset of trace to reproduce the error.
```
thread block = 766,0,0
thread block = 767,0,0
accel-sim.out: abstract_hardware_model.cc:114: void move_warp(warp_inst_t*&, warp_inst_t*&): Assertion `dst->empty()' failed.
Aborted (core dumped)
```
I also have attached the simulation configuration files.
[bert-kernel-2.zip](https://github.com/accel-sim/accel-sim-framework/files/11351402/bert-kernel-2.zip)

**Comment by JRPan** [comment]
#169 Maybe related.

**Comment by William-An** [comment]
Thanks for providing the simulation configuration files!

The issue is caused by the FP unit which accepts both FP16 and FP32 instructions. When an FP32 instruction is issued right after an FP16 instruction, an collision in pipeline register will occur inside the `pipeline_simd_unit`. The reason being the `initiation_interval` is erroneously set to 0 when performing integer division ([link to code](https://github.com/accel-sim/accel-sim-framework/blob/2260456ea5e6a1420f5734f145a4b7d8ab1d4737/gpu-simulator/trace-driven/trace_driven.cc#LL366C1-L375C13)):
```C
    case OP_HADD2:
    case OP_HADD2_32I:
    case OP_HFMA2:
    case OP_HFMA2_32I:
    case OP_HMUL2_32I:
    case OP_HSET2:
    case OP_HSETP2:
      initiation_interval =
          initiation_interval / 2;  
      break;
```

Specifically for your kernel trace file, the two instructions are:
```
0bc0 ffffffff 1 R9 HADD2.F32 2 R255 R4 0 
...
0d10 00000000 1 R9 FMUL.FTZ 2 R9 R8 0 
```

When the `HADD2.F32` instruction is issued, it got assigned to  `m_pipeline_reg[2]` which is moved into `m_pipeline_reg[1]` right before `FMUL.FTZ` is trying to issue to `m_pipeline_reg[1]`, thus causing the collision. A PR is being drafted right now to fix this issue.

**Comment by William-An** [comment]
GPGPU-Sim side is fixed by [gpgpusim/PR30](https://github.com/accel-sim/gpgpu-sim_distribution/pull/30)

**Comment by mahmoodn** [comment]
Thanks for the update. However, if you continue with the same workload and configuration I sent, it ends up with this error
```
thread block = 986181,0,0
thread block = 2266,0,0
thread block = 592618,0,0
thread block = 395436,0,0
thread block = 592619,0,0
thread block = 788659,0,0
thread block = 2267,0,0
accel-sim.out: ../abstract_hardware_model.h:1532: warp_inst_t** register_set::get_free(): Assertion `0 && "No free registers found"' failed.
```
I guess it is somehow related to the modifications for ` m_pipeline_reg`.

**Comment by William-An** [comment]
Hmm weird. I remembered I tested the modifications with your configs and it worked properly. I can double check this later.

**Comment by William-An** [comment]
Okay I pulled the Accel-sim fix (#208) and tested your configs with and without GPGPU-Sim side fix ([GPGPU-SIM/30](https://github.com/accel-sim/gpgpu-sim_distribution/pull/30)) and both runs finished normally.

However I am using the vanilla version of both repo except for the fixes so I have to comment out the last line in your `gpgpusim.config` (`-gpgpu_record_window 1000`).

**Comment by mahmoodn** [comment]
With two fixes, it is now fine. Thank you.
